File: 0045-jump-game-ii/0045-jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& arr) {
         int n = arr.size();
        
        int maxreach = arr[0] ,  step = arr[0] , jump = 1;
        for(int i = 1 ; i < n ; i++){
            if(i == n-1)return jump;
            maxreach = max(maxreach , arr[i] + i);
            step--;
            if(step == 0){
                jump++;
               
                step = maxreach - i;
            }
        }
        return 0;
    }
};

File: 0045-jump-game-ii/0045-jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        int n = nums.length;
        int maxreach = 0 , jump = 0 , currentend = 0 ;
        for(int i = 0 ; i < n-1 ; i++){
            maxreach = Math.max(maxreach , i + nums[i]);
            if(i == currentend){
                jump++;
                currentend = maxreach;
                if(currentend >= n-1)break;
            }
        } 
        return jump;
    }
}